src/ebmc/transition_system.cpp

```diff
@@ -0,0 +1,246 @@
+
+  if(!infile)
+  {
+    message.error() << "failed to open input file `" << filename << "'"
+                    << messaget::eom;
+    return true;
+  }
+
+  auto &lf = language_files.add_file(filename);
+  lf.filename = filename;
+  lf.language = get_language_from_filename(filename);
+
+  if(lf.language == nullptr)
+  {
+    source_locationt location;
+    location.set_file(filename);
+    message.error().source_location = location;
+    message.error() << "failed to figure out type of file" << messaget::eom;
+    return true;
+  }
+
+  languaget &language = *lf.language;
+  language.set_message_handler(message_handler);
+
+  message.status() << "Parsing " << filename << messaget::eom;
+
+  if(language.parse(infile, filename))
+  {
+    message.error() << "PARSING ERROR\n";
+    return true;
+  }
+
+  lf.get_modules();
+
+  return false;
+}
+
+bool parse(
+  const cmdlinet &cmdline,
+  language_filest &language_files,
+  message_handlert &message_handler)
+{
+  for(unsigned i = 0; i < cmdline.args.size(); i++)
+  {
+    if(parse(cmdline.args[i], language_files, message_handler))
+      return true;
+  }
+  return false;
+}
+
+bool get_main(
+  const cmdlinet &cmdline,
+  message_handlert &message_handler,
+  transition_systemt &transition_system)
+{
+  std::string top_module;
+
+  if(cmdline.isset("module"))
+    top_module = cmdline.get_value("module");
+  else if(cmdline.isset("top"))
+    top_module = cmdline.get_value("top");
+
+  try
+  {
+    transition_system.main_symbol =
+      &get_module(transition_system.symbol_table, top_module, message_handler);
+    transition_system.trans_expr =
+      to_trans_expr(transition_system.main_symbol->value);
+  }
+
+  catch(int e)
+  {
+    return true;
+  }
+
+  return false;
+}
+
+void make_next_state(exprt &expr)
+{
+  for(auto &sub_expression : expr.operands())
+    make_next_state(sub_expression);
+
+  if(expr.id() == ID_symbol)
+    expr.id(ID_next_symbol);
+}
+
+int get_transition_system(
+  const cmdlinet &cmdline,
+  ui_message_handlert &message_handler,
+  transition_systemt &transition_system)
+{
+  messaget message(message_handler);
+
+  // do -I
+  if(cmdline.isset('I'))
+    config.verilog.include_paths = cmdline.get_values('I');
+
+  if(cmdline.isset("preprocess"))
+    return preprocess(cmdline, message_handler);
+
+  //
+  // parsing
+  //
+  language_filest language_files;
+
+  if(parse(cmdline, language_files, message_handler))
+    return 1;
+
+  if(cmdline.isset("show-parse"))
+  {
+    language_files.show_parse(std::cout);
+    return 0;
+  }
+
+  //
+  // type checking
+  //
+
+  message.status() << "Converting" << messaget::eom;
+
+  if(language_files.typecheck(transition_system.symbol_table, message_handler))
+  {
+    message.error() << "CONVERSION ERROR" << messaget::eom;
+    return 2;
+  }
+
+  if(cmdline.isset("show-modules"))
+  {
+    show_modules(transition_system.symbol_table, message_handler.get_ui());
+    return 0;
+  }
```

Comment:
Is this the only reason we need a `ui_message_handlert` here? Perhaps this should be factored out, or else various other places should own a reference to a `ui_message_handlert` instead of a `message_handlert` to avoid the (repeatedly occurring) `static_cast`.

Reply:
Yes, exactly working towards getting rid of `ui_message_handlert` -- coming!

Reply:
#127